packages/fxa-auth-server/lib/senders/templates/index.js

```diff
@@ -4,93 +4,115 @@
 
 'use strict';
 
+const LIB_DIR = '../..';
+
+const error = require(`${LIB_DIR}/error`);
+const fs = require('fs');
+const handlebars = {
+  html: require('handlebars').create(),
+  txt: require('handlebars').create(),
+};
 const path = require('path');
-const P = require('bluebird');
-const handlebars = require('handlebars');
-const readFile = P.promisify(require('fs').readFile);
+const Promise = require(`${LIB_DIR}/promise`);
+
+const readDir = Promise.promisify(fs.readdir);
+const readFile = Promise.promisify(fs.readFile);
+
+const TEMPLATE_FILE = /(.+)\.(html|txt)$/;
+const TEMPLATES_DIR = __dirname;
+const LAYOUTS_DIR = path.join(TEMPLATES_DIR, 'layouts');
+const PARTIALS_DIR = path.join(TEMPLATES_DIR, 'partials');
+
+module.exports = init;
+
+async function init(log) {
+  if (!log) {
+    throw error.unexpectedError();
+  }
+
+  handlebars.html.registerHelper('t', translate);
+  handlebars.txt.registerHelper('t', translate);
+
+  await forEachTemplate(PARTIALS_DIR, (template, name, type) => {
+    handlebars[type].registerPartial(name, template);
+  });
+
+  const templates = new Map();
+  const layouts = new Map();
+
+  await Promise.all([
+    forEachTemplate(TEMPLATES_DIR, compile(templates)),
+    forEachTemplate(LAYOUTS_DIR, compile(layouts)),
+  ]);
+
+  return { render };
 
-handlebars.registerHelper('t', function(string) {
-  if (this.translator) {
-    return this.translator.format(this.translator.gettext(string), this);
+  function translate(string) {
+    if (this.translator) {
+      return this.translator.format(this.translator.gettext(string), this);
+    }
+
+    return string;
   }
-  return string;
-});
 
-function generateTemplateName(str) {
-  if (/^sms\.[A-Za-z]+/.test(str)) {
-    return str;
+  function render(templateName, layoutName, data) {
+    const template = templates.get(templateName);
+
+    if (!template) {
+      log.error('templates.render.invalidArg', {
+        templateName,
+        data,
+      });
+      throw error.unexpectedError();
+    }
+
+    const layout = layouts.get(layoutName) || {};
+    let html, text;
+
+    if (template.html) {
+      html = renderWithOptionalLayout(template.html, layout.html, data);
```

Comment:
This `renderWithOptionalLayout` is a slight change to the old `subscription-templates` behaviour however. It's here because the SMS template doesn't have a layout, and the old code used to bailout if no layout was specified. Now it just renders the content template directly instead.